cli/printlog/compact/compact.go

```diff
@@ -0,0 +1,411 @@
+package compact
+
+import (
+	"bufio"
+	"container/heap"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/cli/log"
+	"github.com/klauspost/compress/zstd"
+	"google.golang.org/protobuf/encoding/protodelim"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	spb "github.com/buildbuddy-io/buildbuddy/proto/spawn"
+)
+
+func printSpawnExec(s *spb.SpawnExec) error {
+	b, err := protojson.MarshalOptions{Multiline: true}.Marshal(s)
+	if err != nil {
+		return fmt.Errorf("failed to marshal remote gRPC log entry: %s", err)
+	}
+	if _, err := os.Stdout.Write(append(b, []byte{'\n'}...)); err != nil {
+		return fmt.Errorf("failed to write to stdout: %s", err)
+	}
+	return nil
+}
+
+func PrintCompactExecLog(path string, sort bool) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	r, err := zstd.NewReader(f)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	slr := NewSpawnLogReconstructor(r)
+
+	var spawns []*spb.SpawnExec
+	for {
+		s, err := slr.GetSpawnExec(sort)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed getting spawn exec: %s", err)
+		}
+		if sort {
+			spawns = append(spawns, s)
+			continue
+		}
+
+		if err := printSpawnExec(s); err != nil {
+			return err
+		}
+	}
+	if !sort {
+		return nil
+	}
+
+	// TODO(sluongng): move StableSort to separate func and write tests
+
+	outputToSpawns := make(map[string][]*spb.SpawnExec)
+	for _, s := range spawns {
+		for _, o := range s.GetActualOutputs() {
+			ss, ok := outputToSpawns[o.GetPath()]
+			if !ok {
+				outputToSpawns[o.GetPath()] = []*spb.SpawnExec{s}
+				continue
+			}
+			outputToSpawns[o.GetPath()] = append(ss, s)
+		}
+	}
+
+	blockedBy := newIdentitySetMultiMap[*spb.SpawnExec, *spb.SpawnExec]()
+	blocking := newIdentitySetMultiMap[*spb.SpawnExec, *spb.SpawnExec]()
+
+	queue := make(priorityQueue, len(spawns))
+
+	for _, s := range spawns {
+		if s == nil {
+			// TODO: why nil?
+			continue
+		}
+		blocked := false
+		for _, input := range s.GetInputs() {
+			for _, blocker := range outputToSpawns[input.GetPath()] {
+				blockedBy.put(s, blocker)
+				blocking.put(blocker, s)
+				blocked = true
+			}
+		}
+		if blocked {
```

Comment:
```suggestion
		if !blocked {
```
otherwise you are dropping all spawns that aren't blocked